velox/functions/prestosql/URLFunctions.h

```diff
@@ -0,0 +1,281 @@
+  return StringView(sub.first, sub.length());
+}
+
+template <typename TInString>
+bool UrlParser::parse(TInString rawUrl, boost::cmatch& match) {
+  static const boost::regex uriRegex(
+      "([a-zA-Z][a-zA-Z0-9+.-]*):" // scheme:
+      "([^?#]*)" // authority and path
+      "(?:\\?([^#]*))?" // ?query
+      "(?:#(.*))?"); // #fragment
+
+  if (UNLIKELY(!boost::regex_match(rawUrl.data(), match, uriRegex))) {
+    return false;
+  }
+
+  return true;
+}
+
+template <typename T>
+struct UrlExtractProtocolFunction : UrlParser {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // Results refer to strings in the first argument.
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& urlString) {
+    boost::cmatch match;
+    if (!parse(urlString, match)) {
+      result.setEmpty();
+    } else {
+      auto scheme = submatch(match, 1);
+      result.setNoCopy(scheme);
+    }
+    return true;
+  }
+};
+
+template <typename T>
+struct UrlExtractFragmentFunction : UrlParser {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // Results refer to strings in the first argument.
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& urlString) {
+    boost::cmatch match;
+    if (!parse(urlString, match)) {
+      result.setEmpty();
+    } else {
+      auto scheme = submatch(match, 4);
+      result.setNoCopy(scheme);
+    }
+    return true;
+  }
+};
+
+template <typename T>
+struct UrlExtractHostFunction : UrlParser {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // Results refer to strings in the first argument.
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& urlString) {
+    boost::cmatch match;
+    if (parse(urlString, match)) {
+      boost::cmatch authAndPathMatch;
+      boost::cmatch authorityMatch;
+      bool hasAuthority;
+
+      if (authorityAndPathMatch(
+              match, authAndPathMatch, authorityMatch, hasAuthority) &&
+          hasAuthority) {
+        result.setNoCopy(submatch(authorityMatch, 3));
+      } else {
+        result.setEmpty();
+      }
+
+      return true;
+    }
+
+    result.setEmpty();
+    return true;
+  }
+};
+
+template <typename T>
+struct UrlExtractPortFunction : UrlParser {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<int64_t>& result,
+      const arg_type<Varchar>& urlString) {
+    boost::cmatch match;
+    result = 0;
```

Comment:
This assignment is not needed, is it?

Reply:
I am not sure if result will be zero set, so I assigned 0 to make sure. Ports cant have 0 as value.

Reply:
This assignment is not needed. The result will either be set to the port extracted from the URL or the function will return null. Let's remove this.